### PR TITLE
Add tests and setup codecov

### PR DESCRIPTION
**Pull request by alecpm**
This branch adds some tests to Griduniverse. Currently they increase coverage from 48% to 62%. The new tests are mostly coverage oriented, rather than aimed at testing specific behaviors, and are a bit monolithic as a result. I expect they will be refactored to be more modular over time as functionality changes and is added which impacts the behavior of the tested methods.

### CODE REVIEW

File: test/test_griduniverse.py [hunk 1]
```diff
@@ -53,6 +52,253 @@ def test_environment_uses_experiments_networks(self, exp):
...
+        exp.grid.players = {'1': mock.Mock()}
+        exp.grid.players['1'].score = 10
+        exp.grid.contagion = 1
+        exp.grid.tax = 1.0
+        exp.grid.food_locations = []
+        exp.grid.frequency_dependence = 1
+        exp.grid.frequency_dependent_payoff_rate = 0
+        exp.socket_session = mock.Mock()
+        exp.publish = mock.Mock()
+        with mock.patch('gevent.sleep') as _fake_sleep:
```
[the remark this answers is not in the document]
Comment (alecpm) [reply test/test_griduniverse.py]: I initially verified the number of calls to `sleep` but then realized that was absurd.

File: test/test_griduniverse.py [hunk 2]
```diff
@@ -53,6 +52,251 @@ def test_environment_uses_experiments_networks(self, exp):
...
+        # These are generally costs for things enabled above
+        for param in dependent_numeric_params:
+            # increment the cost
+            setattr(gridworld, param, getattr(gridworld, param, 0) + 1)
+            assert gridworld.instructions() != instructions, (
+                '{} = {} did not change instructions'.format(
+                    param, getattr(gridworld, param))
+            )
+            instructions = gridworld.instructions()
+
```
[the remark this answers is not in the document]
Comment (alecpm) [reply test/test_griduniverse.py]: Phew!

**Comment by alecpm** [comment]
I've separated out the fixtures you suggested, and factored out a `TestGameLoops` class for `game_loop` and `send_state_thread` tests.

**Comment by alecpm** [comment]
`game_loop` will loop 3 times and `send_state_thread` will loop 4 times because of the different way they do their variable checking. `Gridworld.game_over` and `Gridworld.game_started` will each return False 3 times before returning True.

I kind of hate writing `loop_3x_exp` a whole bunch of times, just because it's kind of ugly, or having to rename it in each test, but it would be clearer.

**Comment by alecpm** [comment]
I'm going to do a rebase anyway to get rid of this dumb patch coverage issue, so I'll do a rename along with that.

**Comment by jessesnyder** [comment]
`loops_3x`? :)

**Comment by alecpm** [comment]
`loop_exp_3x` which gets renamed to `exp` in every test.

**Pull request merged**